### DISCUSSION TITLE
[DX12][WIN32] Unhandled ImGui exception when EnableDebugLayer

### DISCUSSION THREAD

**Discussion by bboczula** [ocornut/imgui#6288] [New Users Build/Link/Run issues ONLY!]
Hello,

I've integrated ImGui to my engine, and was able to show a demo window on top of my rendering, so that was great. Unfortunately, when I switched to Debug configuration, it stardet to crash with "Unhandled Exception" error.

After a little digging, it turned out that this behavior is cause by calling "EnableDebugLayer()" function, even in the Release configuration.

The crash happens with "Exception Unhandled" in _imgui_impl_dx12.cpp_, line 276:
`ctx->SetGraphicsRootDescriptorTable(1, texture_handle);`

I'm also getting this error from DX12:
`D3D12 ERROR: CGraphicsCommandList::SetGraphicsRootDescriptorTable: Specified GPU descriptor handle ptr=0x41 does not refer to a location in a descriptor heap.  [ EXECUTION ERROR #646: INVALID_DESCRIPTOR_HANDLE]`

This is the rendering code:

```
void Sapphire::ImGuiPass::Render(DX12CommandList* commandList, RenderInterface* renderInterface, std::vector<GameObject*> objects)
{
	PIXBeginEvent(commandList->GetCommandList(), PIX_COLOR(255, 255, 255), "ImGuiPass");

	ImGui_ImplDX12_NewFrame();
	ImGui_ImplWin32_NewFrame();

	ImGui::NewFrame();
	ImGui::ShowDemoWindow();
	ImGui::Render();
	ImGui::EndFrame();

	ImGui_ImplDX12_RenderDrawData(ImGui::GetDrawData(), renderInterface->GetCommandList()->GetCommandList());

	PIXEndEvent(commandList->GetCommandList());
}
```

Regarding initalization, I simply allocate one descriptor on my global SRV_CBV_UAV descriptor heap that is GPU-visible and provide that to ImGui initialization functions.

```
auto descriptorIndex = renderInterface->GetSrvDescriptorHeap()->GetHeapSize();
D3D12_CPU_DESCRIPTOR_HANDLE srvHandle;
srvHandle.ptr = renderInterface->GetSrvDescriptorHeap()->AllocateDescriptor();

D3D12_GPU_DESCRIPTOR_HANDLE gpuSrvHandle;
gpuSrvHandle.ptr = renderInterface->GetSrvDescriptorHeap()->GetGpuDescriptor(descriptorIndex);

ImGui_ImplWin32_Init(renderInterface->GetHwnd());
ImGui_ImplDX12_Init(renderInterface->GetDevice()->GetDevice(), 2, DXGI_FORMAT_R8G8B8A8_UNORM,
	nullptr, srvHandle, gpuSrvHandle);
```

I've passed `nullptr` as ID3D12DescriptorHeap, because it doesn't seem to be used anywhere, but even with a working pointer I still get the same error.

Does anyone else had this problem maybe?

**Comment by bboczula** [comment]
Ok, this was my mistake, in order to get the GPU descriptor, I was calling `GetCPUDescriptorHandleForHeapStart` instead of `GetGPUDescriptorHandleForHeapStart`.

**Reply by ravo10** [reply]
I had forgotten to release my ID3D12Resources...! E.g.:

```
...
if (texture1) { texture1->Release(); texture1= nullptr; }
if (texture2) { texture2->Release(); texture2= nullptr; }

CleanupDeviceD3D();
```
